python/GafferTest/EditScopeTest.py

```diff
@@ -0,0 +1,134 @@
+#        disclaimer in the documentation and/or other materials provided with
+#        the distribution.
+#
+#      * Neither the name of John Haddon nor the names of
+#        any other contributors to this software may be used to endorse or
+#        promote products derived from this software without specific prior
+#        written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+##########################################################################
+
+import unittest
+
+import Gaffer
+import GafferTest
+
+class EditScopeTest( GafferTest.TestCase ) :
+
+	def setUp( self ) :
+
+		GafferTest.TestCase.setUp( self )
+
+		def createIntProcessor() :
+
+			n = Gaffer.ContextVariables()
+			n.setup( Gaffer.IntPlug() )
+			return n
+
+		Gaffer.EditScope.registerProcessor( "Test", createIntProcessor )
+
+	def tearDown( self ) :
+
+		GafferTest.TestCase.tearDown( self )
+
+		Gaffer.EditScope.deregisterProcessor( "Test" )
+
+	def testProcessorRegistry( self ) :
+
+		self.assertIn( "Test", Gaffer.EditScope.registeredProcessors() )
+
+		Gaffer.EditScope.registerProcessor( "Test", lambda : Gaffer.Node() )
+		self.assertEqual( Gaffer.EditScope.registeredProcessors().count( "Test" ), 1 )
+
+		Gaffer.EditScope.deregisterProcessor( "Test" )
+		self.assertEqual( Gaffer.EditScope.registeredProcessors().count( "Test" ), 0 )
+
+	def testSetup( self ) :
+
+		e = Gaffer.EditScope()
+		self.assertNotIn( "in", e )
+		self.assertNotIn( "out", e )
+
+		e.setup( Gaffer.IntPlug() )
+
+		self.assertIn( "in", e )
+		self.assertIn( "out", e )
+
+		self.assertIsInstance( e["in"], Gaffer.IntPlug )
+		self.assertIsInstance( e["out"], Gaffer.IntPlug )
+		self.assertEqual( e["out"].source(), e["in"] )
+
+		self.assertEqual( e.correspondingInput( e["out"] ), e["in"] )
+
+	def testAcquireProcessor( self ) :
+
+		e = Gaffer.EditScope()
+		e.setup( Gaffer.IntPlug() )
+
+		self.assertIsNone( e.acquireProcessor( "Test", createIfNecessary = False ) )
+		p = e.acquireProcessor( "Test" )
+		self.assertIsInstance( p, Gaffer.ContextVariables )
+		self.assertEqual( e["out"].source(), p["out"] )
+		self.assertEqual( p["in"].source(), e["in"] )
+
+		self.assertEqual( p, e.acquireProcessor( "Test" ) )
+
```

Comment:
Tested messed up internal connections in ab6fbf8. I've avoided testing insertion position for now, because we don't yet have examples of processors which need to be inserted in a particular processor with respect to another - I was going to add the API for that when we had an actual example.

Reply:
Cool thanks, wondered if we wanted to also test a broken edit scope where there is no route back to `in`?

Reply:
Added `testBrokenInternalConnection()` in https://github.com/GafferHQ/gaffer/pull/3666/commits/94da9ff0aeb115a5c2351651355232d098b3ed5a.